### PR TITLE
image lightness filter

### PR DESCRIPTION
**Pull request by ip75**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
issue #998

Create a lightness filter

### CODE REVIEW

**Comment by ip75** [comment]
Thank you for reminding. I see all messages of CI. I'll fix it while this
week I hope. Working hard and urgent domestic matters.

**Comment by ip75** [comment]
dotnet :     WorksWithAllBitDepths<Rgba32>(provider: TestPattern24x24[Rgba32], pngColorType: Palette, pngBitDepth: Bit8) [FAIL]
At line:1 char:1
+ dotnet xunit -nobuild -c Release -framework net462 -x86
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (    WorksWithAl...h: Bit8) [FAIL]:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
      System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown.
      Stack Trace:


I do not understand what does it mean .

**Comment by tocsoft** [comment]
@ip75  The issue with the CLA is that you created commits with 2 different email addresses one associated with your github account and one not. You will either need to fix your commits (just squashing them all down and force pushing would do) ensuring you use the email address associated with your github account as the commit author.

File: src/ImageSharp/Processing/KnownFilterMatrices.cs [hunk 1]
```diff
@@ -445,22 +445,19 @@ public static ColorMatrix CreateLightnessFilter(float amount)
...
-            ///
+            // James Jackson-South @JimBobSquarePants 03:54
+            // Our colormatrix is a column-major version of the Android colormatrix
+            //
+            //     | 0| 1| 2| 3| 4|   |0|5|10|15|   |M11|M12|M13|M14|
+            //     | 5| 6| 7| 8| 9|   |1|6|11|16|   |M21|M22|M23|M24|
+            //     |10|11|12|13|14| = |2|7|12|17| = |M31|M32|M33|M34|
+            //     |15|16|17|18|19|   |3|8|13|18|   |M41|M42|M43|M44|
+            //                        |4|9|14|19|   |M51|M52|M53|M54|
+            // James Jackson-South @JimBobSquarePants 03:54
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/KnownFilterMatrices.cs]: Delete these additional comments please, they are just noise.

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+using SixLabors.ImageSharp.Processing.Processors.Filters;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing
+{
+    /// <summary>
+    /// Defines extensions that allow the alteration of the hue component of an <see cref="Image"/>
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: Class and value descriptions are copy pasted from hue and will need updating.

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    /// using Mutate/Clone.
+    /// </summary>
+    public static class LightnessExtension
+    {
+        /// <summary>
+        /// Alters the hue component of the image.
+        /// </summary>
+        /// <param name="source">The image this method extends.</param>
+        /// <param name="lightness">Lightness parameter of image in HSL color scheme.</param>
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: I'm not sure this is an accurate description. 

In HSL lightness corresponds to the arithmetic mean of the largest and smallest RGB components, In a color matrix the W(hiteness) in RGBAW is basically added to multipliers of each RGB component.

File: src/ImageSharp/Processing/KnownFilterMatrices.cs [hunk 4]
```diff
@@ -432,6 +432,30 @@ public static ColorMatrix CreateSaturateFilter(float amount)
...
+        /// <remarks>
+        /// A value of 0 will create an image that is completely black. A value of 1 makes the image completely white.
+        /// Other values are linear multipliers on the effect. Values of an amount over 1 are allowed, providing brighter results.
+        /// </remarks>
+        /// <param name="amount">The proportion of the conversion. Must be greater than or equal to 0.</param>
+        /// <returns>The <see cref="ColorMatrix"/></returns>
+        public static ColorMatrix CreateLightnessFilter(float amount)
+        {
+            Guard.MustBeBetweenOrEqualTo(amount, 0, 1F, nameof(amount));
+            return new ColorMatrix
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/KnownFilterMatrices.cs]: We're going to need some tests for this. I recommend copying the brightness ones.

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs (hunk 2 above)
Comment (ip75) [reply src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: you right. i copied from hue extension and was not able to change all text in a rush.

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs (hunk 3 above)
Comment (ip75) [reply src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: image

File: src/ImageSharp/Processing/KnownFilterMatrices.cs (hunk 4 above)
Comment (ip75) [reply src/ImageSharp/Processing/KnownFilterMatrices.cs]: I'll try to do it.

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: What I mean is that we are not operating within the HSL colorspace. HSL lightness would likely require additional calculation.

Lightness can mean many different things. The algorithm I supplied you gave a non-weighted alteration to the white value of RGBA. 

Take a look at this article to help you understand the difference.
https://thoughtbot.com/blog/closer-look-color-lightness

File: tests/ImageSharp.Tests/Processing/Processors/Filters/LightnessTest.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    [GroupOutput("Filters")]
+    public class LightnessTest
+    {
+        private readonly ImageComparer imageComparer = ImageComparer.Tolerant(0.007F);
+
+        public static readonly TheoryData<float> LightnessValues
+        = new TheoryData<float>
+        {
+            .5F,
+            .5F
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Processing/Processors/Filters/LightnessTest.cs]: This is the same value twice. Why can it not be greater than 1?

File: src/ImageSharp/Processing/Extensions/LightnessExtension.cs (hunk 3 above)
Comment (ip75) [reply src/ImageSharp/Processing/Extensions/LightnessExtension.cs]: Ok. I'll remove all hsl references.

File: tests/ImageSharp.Tests/Processing/Processors/Filters/LightnessTest.cs (hunk 5 above)
Comment (ip75) [reply tests/ImageSharp.Tests/Processing/Processors/Filters/LightnessTest.cs]: > 1 is the same as 1. Image will be white 100%

File: src/ImageSharp/Processing/KnownFilterMatrices.cs (hunk 1 above)
Comment (ip75) [reply src/ImageSharp/Processing/KnownFilterMatrices.cs]: I've done it.

**Comment by JimBobSquarePants** [comment]
Gave this a once over and updated the reference image repo. Should be good to merge once built.

**Comment by ip75** [comment]
million errors while compiling sources.

...............

1>ColorSpaces\Conversion\ColorSpaceConverter.CieLchuv.cs(371,13,371,18): error CS0122: 'Guard' is inaccessible due to its protection level
1>ColorSpaces\Conversion\ColorSpaceConverter.CieLchuv.cs(404,13,404,18): error CS0122: 'Guard' is inaccessible due to its protection level
1>ColorSpaces\Conversion\ColorSpaceConverter.CieLchuv.cs(436,13,436,18): error CS0122: 'Guard' is inaccessible due to its protection level
1>ColorSpaces\Conversion\ColorSpaceConverter.CieLuv.cs(37,13,37,18): error CS0122: 'Guard' is inaccessible due to its protection level
1>ColorSpaces\Conversion\ColorSpaceConverter.CieLuv.cs(69,13,69,18): error CS0122: 'Guard' is inaccessible due to its protection level

...............

**Comment by JimBobSquarePants** [comment]
Something wrong in your local build. The only error throwing is the difference in Histogram equalization reference images from #1017

**Comment by JimBobSquarePants** [comment]
Try deleting the .vs folder in your local build. Sometimes Visual Studio will cache incorrectly breaking builds

**Comment by ip75** [comment]
image
access level internal in SixLabors.Core.dll but compilled library is ImageSharp

**Comment by ip75** [comment]
I have not retrieved submodules from project.

**Comment by ip75** [comment]
James, thank you very much for your help. :)

**Comment by JimBobSquarePants** [comment]
My pleasure, great suggestion!

**Pull request merged**